dulwich/porcelain.py

```diff
@@ -1210,11 +1210,14 @@ def status(repo=".", ignored=False):
     with open_repo_closing(repo) as r:
         # 1. Get status of staged
         tracked_changes = get_tree_changes(r)
+        for key in tracked_changes.keys():
+            tracked_changes[key] = [file.decode() for file in tracked_changes[key]]
         # 2. Get status of unstaged
         index = r.open_index()
         normalizer = r.get_blob_normalizer()
         filter_callback = normalizer.checkin_normalize
         unstaged_changes = list(get_unstaged_changes(index, r.path, filter_callback))
+        unstaged_changes = [file.decode() for file in unstaged_changes]
```

Comment:
This is going to decode with the current encoding (e.g. utf-8) which may not work if the filename contains bytes that can't be decoded with the current encoding (e.g. utf-8) or even if they can it may not result in the string you were expecting (file system encoding and display encoding may be different).

Reply:
Can we set the default encoding in the script which using dulwich to resolve it?

Reply:
That would be inconsistent with the way the rest of Dulwich works. And if you're making the caller worry about the encoding, why not let them do the decoding as well?

Reply:
For some uses cases, bytes are sufficient. You also can choose what to do in that case if you want to display the output - you can ignore invalid characters, replace them with surrogates or e.g. use chardet to try to figure out the proper encoding if it isn't utf8.

Reply:
I'd be open to returning a surrogateescape-decoded string here, since it's possible for the caller to undo that and retrieve the original bytes but that would still be a breaking change for existing users of the API. We should also force utf8 in that case (since that's the convention for git repositories) rather than whatever the users' encoding is.

Reply:
I think the function in porcelain should work like the git subcommands(a higher function), or I should decode the bytes after using `porcelain.status`.